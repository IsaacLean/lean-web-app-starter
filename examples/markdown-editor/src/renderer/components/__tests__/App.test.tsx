import { getByText, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import App from '../App'

beforeEach(() => {
    // Mock the API
    window.api = {
        checkForMdChange: jest.fn(
            // Markdown changes will always be boolean
            () => new Promise((resolve) => resolve(true)),
        ),
        checkForUnsavedChanges: jest.fn(),
        /* eslint-disable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-empty-function */
        // @ts-ignore
        onColorModeMenu: jest.fn(() => () => {}),
        // @ts-ignore
        onFocusModeToggle: jest.fn(() => () => {}),
        // @ts-ignore
        onMainErrorMessage: jest.fn(() => () => {}),
        // @ts-ignore
        onMainHtmlExportDialog: jest.fn(() => () => {}),
        // @ts-ignore
        onMainMarkdownOpenDialog: jest.fn(() => () => {}),
        // @ts-ignore
        onMainSaveFile: jest.fn(() => () => {}),
        // @ts-ignore
        onOpenFileSuccess: jest.fn(() => () => {}),
        // @ts-ignore
        onSaveFileSuccess: jest.fn(() => () => {}),
        /* eslint-enable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-empty-function */
        saveFile: jest.fn(),
        showFileOpenDialog: jest.fn(),
        showHtmlExportDialog: jest.fn(),
        showInFolder: jest.fn(),
        showUnsavedChangesDialog: jest.fn(),
        syncColorModeMenu: jest.fn(),
    }

    // Allow onbeforeunload to be redefined so Jest doesn't throw errors
    Object.defineProperty(window, 'onbeforeunload', { writable: true })
})

/**
 * Setup user-event and render the component.
 */
const setupTest = () => {
    const user = userEvent.setup()
    const { asFragment } = render(<App />)

    return { user, asFragment }
}

test('App component matches snapshot', () => {
    const { asFragment } = setupTest()
    expect(asFragment()).toMatchSnapshot()
})

test('Markdown in editor generates HTML in preview', async () => {
    expect.assertions(2)

    const { user } = setupTest()

    // Type markdown in the editor
    await screen.findByRole('textbox')
    await user.keyboard('# Hello World{Enter}Foobar!')

    const preview = screen.getByRole('article')

    // Expect the correct HTML to be generated in the preview
    await waitFor(() => {
        expect(
            screen.getByRole('heading', { name: /hello world/i }),
        ).toBeInTheDocument()
    })
    expect(getByText(preview, /foobar!/i)).toBeInTheDocument()
})

test('Top bar hides when focus mode button is clicked', async () => {
    expect.assertions(2)

    const { user } = setupTest()

    // Expect top bar to be visible
    expect(
        screen.getByRole('button', { name: 'Open File' }),
    ).toBeInTheDocument()

    // Click the focus mode button
    await user.click(screen.getByRole('button', { name: 'Focus Mode' }))

    // Expect top bar to be hidden
    expect(
        screen.queryByRole('button', { name: 'Open File' }),
    ).not.toBeInTheDocument()
})
