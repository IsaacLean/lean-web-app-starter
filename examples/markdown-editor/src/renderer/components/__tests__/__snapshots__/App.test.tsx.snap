// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="flex h-screen w-screen flex-col bg-zinc-200 text-zinc-500 subpixel-antialiased dark:bg-gray-950 dark:text-white"
  >
    <nav
      class="flex w-full justify-between border-b border-zinc-300 px-3 py-2 dark:border-gray-700"
    >
      <section
        class="space-x-2"
      >
        <button
          class="btn"
        >
          Open File
        </button>
        <button
          class="btn"
          disabled=""
        >
          Save File
        </button>
        <button
          class="btn"
        >
          Export as HTML
        </button>
        <button
          class="btn"
          disabled=""
        >
          Show In Folder
        </button>
      </section>
      <section
        class="flex space-x-2"
      >
        <button
          class="btn flex space-x-2"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.036 12.322a1.012 1.012 0 0 1 0-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span>
            Focus Mode
          </span>
        </button>
        <button
          class="btn flex space-x-2"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 3v1.5M4.5 8.25H3m18 0h-1.5M4.5 12H3m18 0h-1.5m-15 3.75H3m18 0h-1.5M8.25 19.5V21M12 3v1.5m0 15V21m3.75-18v1.5m0 15V21m-9-1.5h10.5a2.25 2.25 0 0 0 2.25-2.25V6.75a2.25 2.25 0 0 0-2.25-2.25H6.75A2.25 2.25 0 0 0 4.5 6.75v10.5a2.25 2.25 0 0 0 2.25 2.25Zm.75-12h9v9h-9v-9Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span>
            Sys. Pref. Mode
          </span>
        </button>
      </section>
    </nav>
    <section
      class="flex flex-1 flex-row overflow-hidden"
    >
      <form
        class="flex w-1/2 flex-1 flex-col border-r border-zinc-300 dark:border-gray-700"
      >
        <h1
          class="border-b border-zinc-300 px-3 py-1 dark:border-gray-700"
        >
          Editor
        </h1>
        <textarea
          class="form-textarea flex-1 resize-none border-none bg-white font-mono text-xs text-black focus:ring-0 dark:bg-gray-900 dark:text-white "
        />
      </form>
      <section
        class="flex w-1/2 flex-1 flex-col"
      >
        <h1
          class="border-b border-zinc-300 px-3 py-1 dark:border-gray-700"
        >
          Preview
        </h1>
        <article
          class="prose flex-1 overflow-scroll bg-white px-3 py-2 dark:prose-invert dark:bg-gray-900"
          style="text-rendering: optimizeLegibility;"
        />
      </section>
    </section>
  </main>
</DocumentFragment>
`;
